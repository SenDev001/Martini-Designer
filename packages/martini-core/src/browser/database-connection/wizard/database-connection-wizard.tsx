import { SelectionService } from "@theia/core";
import { OpenerService } from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { Field, Form, Formik, FormikErrors, FormikProps } from "formik";
import { inject, injectable } from "inversify";
import { noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import * as Yup from "yup";
import {
    DatabaseType,
    getTypeLabel, MartiniDatabaseConnectionManager
} from "../../../common/database-connection/martini-database-connection-manager";
import { UpDownLabel } from "../../components/up-down-label";
import { FormEffect, FormRow, OnFormChange, validateSchema } from "../../form/form";
import { Progress } from "../../progress/progress-service";
import { AbstractWizard, AbstractWizardPage, Wizard } from "../../wizard/wizard";
import { WizardContribution } from "../../wizard/wizard-contribution";
import { isDbConnectionSelection } from "../database-connection-contribution";
import { DatabaseConnectionEditor } from "../editor/database-connection-editor";

@injectable()
export class DatabaseConnectionWizardContribution implements WizardContribution {
    @inject(SelectionService)
    private selectionService: SelectionService;
    @inject("Factory<DatabaseConnectionWizard>")
    private factory: () => DatabaseConnectionWizard;

    readonly wizardType = "new";
    readonly id = "database-connection";
    readonly label = messages.database_connection;
    readonly description = messages.create_db_connection;
    readonly iconClass = "martini-icon martini-database-icon";
    readonly keybinding = "ctrlcmd+alt+n d";
    readonly primary = true;

    async createWizard(): Promise<Wizard> {
        return this.factory();
    }

    isVisible() {
        const selection = this.selectionService.selection;
        return isDbConnectionSelection(selection) || (selection instanceof Array && selection[0] === "Databases");
    }

}

@injectable()
export class DatabaseConnectionWizardPage extends AbstractWizardPage {
    existingNames: string[] = [];
    private readonly defaultConfig: DatabaseConnectionConfig = {
        type: DatabaseType.JDBC,
        name: "MyConnection"
    };
    config: DatabaseConnectionConfig = this.defaultConfig;

    render(): React.ReactNode {
        return <CreateDatabaseConnectionForm
            existingNames={this.existingNames}
            defaultConfig={this.defaultConfig}
            onValidate={errors => this.handleValidate(errors)}
            onChange={config => this.handleChange(config)}
        />;
    }

    private handleValidate(errors: FormikErrors<DatabaseConnectionConfig>) {
        this.complete = Object.keys(errors).length === 0;
    }

    private handleChange(config: DatabaseConnectionConfig) {
        this.config = config;
    }
}

export interface DatabaseConnectionConfig {
    type: DatabaseType;
    name: string;
}

export interface CreateDatabaseConnectionFormProps {
    existingNames: string[];
    defaultConfig?: DatabaseConnectionConfig;
    onValidate: (errors: FormikErrors<DatabaseConnectionConfig>) => void;
    onChange: (config: DatabaseConnectionConfig) => void;
}

const databaseConnectionNameRegExp = /^[a-zA-Z0-9\\_-]+$/;

export const CreateDatabaseConnectionForm: React.FC<CreateDatabaseConnectionFormProps> = (
    { existingNames, defaultConfig, onValidate, onChange }) => {

    const schema = React.useMemo(() => Yup.object().shape<Omit<DatabaseConnectionConfig, "type">>({
        name: Yup.string()
            .required()
            .max(30)
            .matches(databaseConnectionNameRegExp, messages.only_alphanumeric)
            .test("unique-name", messages.connection_name_exists, value => !existingNames.includes(value!))
    }), []);
    const validate = (config: DatabaseConnectionConfig) => validateSchema(schema, config, onValidate);

    const handleKeyDown = (e: React.KeyboardEvent, props: FormikProps<DatabaseConnectionConfig>) => {
        if (e.key !== "ArrowUp" && e.key !== "ArrowDown")
            return;
        const types = Object.values(DatabaseType);
        let index = types.indexOf(props.values.type);
        if (e.key === "ArrowUp")
            index = index - 1 < 0 ? types.length - 1 : index - 1;
        else if (e.key === "ArrowDown")
            index = index + 1 >= types.length ? 0 : index + 1;
        const type = types[index];
        props.setValues({
            ...props.values,
            type
        });
        e.preventDefault();
    };

    return <Formik
        initialValues={defaultConfig}
        initialTouched={{
            type: true,
            name: true
        }}
        validate={validate}
        validateOnMount={true}
        onSubmit={noop}
    >
        {props => (
            <Form style={{
                display: "grid",
                gridColumnGap: 10,
                gridTemplateColumns: "max-content 1fr"
            }}>
                <FormEffect deps={[existingNames]} />
                <OnFormChange onChange={onChange} />
                <FormRow name="type" label={messages.type}>
                    <Field as="select" name="type" className="theia-select">
                        {Object.values(DatabaseType).map(type => (
                            <option key={type}
                                value={type}>{getTypeLabel(type)}</option>))}
                    </Field>
                </FormRow>
                <FormRow name="name" label={messages.name} gridTemplateColumns="1fr max-content">
                    <Field
                        type="text"
                        name="name"
                        className="theia-input"
                        autoFocus
                        onKeyDown={(e: React.KeyboardEvent) => handleKeyDown(e, props)}
                    />
                    <UpDownLabel toolTip={messages.pressing_updown_changes_type} />
                </FormRow>
            </Form>)}
    </Formik>;
};

@injectable()
export class DatabaseConnectionWizard extends AbstractWizard {
    readonly title = messages.create_db_connection_title;

    constructor(
        @inject(MartiniDatabaseConnectionManager)
        private readonly dbConnectionMngr: MartiniDatabaseConnectionManager,
        @inject(DatabaseConnectionWizardPage)
        private readonly page: DatabaseConnectionWizardPage,
        @inject(OpenerService)
        private readonly openerService: OpenerService) {
        super();
        this.pages.push(page);
    }

    async init(progress: Progress): Promise<void> {
        progress.report({
            message: messages.fetching_database_connections
        });
        const connections = await this.dbConnectionMngr.getAll();
        this.page.existingNames = connections.map(c => c.name);
        return super.init(progress);
    }

    async performFinish(): Promise<boolean> {
        const uri: URI = new URI(`${DatabaseConnectionEditor.URI_SCHEME}://${encodeURI(this.page.config.name)}`
            + `?type=${this.page.config.type}&new=true`);
        const handler = await this.openerService.getOpener(uri);
        handler.open(uri);
        return true;
    }

}
