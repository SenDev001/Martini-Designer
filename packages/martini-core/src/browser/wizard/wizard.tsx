import { Disposable, DisposableCollection, Emitter, Event } from "@theia/core";
import { injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import { BaseDialog } from "../dialogs/dialogs";
import { Progress } from "../progress/progress-service";

export interface Wizard extends Disposable {
    currentPage: WizardPage | undefined;
    firstPage: WizardPage;
    title: string;
    onCompleteChanged: Event<void>;
    onUpdate: Event<void>;
    hasNext: boolean;
    hasPrevious: boolean;
    initialSize?: { width?: number, height?: number; };

    init(progress: Progress): Promise<void>;

    performFinish(progress: Progress): Promise<boolean>;

    canFinish(): boolean;

    getNextPage(current: WizardPage | undefined): WizardPage | undefined;

    getPreviousPage(current: WizardPage | undefined): WizardPage | undefined;

    render(): React.ReactNode;

    next(): void;

    previous(): void;
}

export interface WizardPage extends Disposable {
    description?: string;
    complete: boolean;
    onCompleteChanged: Event<void>;
    onUpdate: Event<void>;

    render(): React.ReactNode;

    dispose(): void;
}

@injectable()
export abstract class AbstractWizard implements Wizard {
    protected onCompleteChangedEmitter = new Emitter<void>();
    readonly onCompleteChanged = this.onCompleteChangedEmitter.event;
    protected onUpdateEmitter = new Emitter<void>();
    readonly onUpdate = this.onUpdateEmitter.event;

    abstract title: string;
    protected pages: WizardPage[] = [];
    protected history: WizardPage[] = [];
    protected current: WizardPage;
    protected toDispose = new DisposableCollection();

    /**
     * Initializes the wizard. If this method is overridden it should call super when all the pages have been added already.
     *
     * @param progress used to report progress
     */
    async init(progress: Progress): Promise<void> {
        if (this.pages.length) this.current = this.firstPage;
        this.toDispose.pushAll(this.pages);
    }

    abstract performFinish(progress: Progress): Promise<boolean>;

    canFinish(): boolean {
        return (
            this.pages.length === 0 ||
            (this.currentPage !== undefined &&
                this.currentPage.complete &&
                this.getNextPage(this.current) === undefined)
        );
    }

    getNextPage(current: WizardPage | undefined): WizardPage | undefined {
        if (!current) return undefined;
        const index = this.pages.indexOf(current);
        if (index >= 0 && index + 1 < this.pages.length)
            return this.pages[index + 1];
        return undefined;
    }

    getPreviousPage(current: WizardPage | undefined): WizardPage | undefined {
        return this.history.length === 0 ? undefined : this.history[this.history.length - 1];
    }

    render(): React.ReactNode {
        const page = this.currentPage;
        if (page) {
            return (
                <div style={{ display: "grid", overflowY: "hidden" }}>
                    {page.description && (
                        <>
                            <div
                                style={{
                                    marginBottom: 5,
                                    padding: "6px 12px"
                                }}
                            >
                                {page.description}
                            </div>
                            <div
                                style={{
                                    borderTop: "1px solid var(--theia-panel-border)",
                                    width: "100%",
                                    marginBottom: 5
                                }}
                            />
                        </>
                    )}
                    <div style={{ padding: "6px 12px", overflowY: "auto" }}>{page.render()}</div>
                </div>
            );
        }
        return <></>;
    }

    get firstPage() {
        return this.pages[0];
    }

    get currentPage() {
        return this.current;
    }

    get hasNext() {
        return this.pages.length > 1;
    }

    get hasPrevious() {
        return this.hasNext;
    }

    next(): void {
        if (this.current && this.current.complete) {
            const next = this.getNextPage(this.current);
            if (next) {
                this.history.push(this.current);
                this.current = next;
                this.onCompleteChangedEmitter.fire();
                this.onUpdateEmitter.fire();
            }
        }
    }

    previous(): void {
        if (this.current) {
            const previous = this.getPreviousPage(this.current);
            if (previous) {
                this.current = previous;
                this.history.pop();
                this.onCompleteChangedEmitter.fire();
                this.onUpdateEmitter.fire();
            }
        }
    }

    dispose(): void {
        this.toDispose.dispose();
    }
}

@injectable()
export abstract class AbstractWizardPage implements WizardPage {
    protected _complete = false;
    protected onCompleteChangedEmitter = new Emitter<void>();
    readonly onCompleteChanged = this.onCompleteChangedEmitter.event;
    protected onUpdateEmitter = new Emitter<void>();
    readonly onUpdate = this.onUpdateEmitter.event;
    protected toDispose = new DisposableCollection();
    protected _description = "";

    abstract render(): React.ReactNode;

    get complete() {
        return this._complete;
    }

    set complete(complete: boolean) {
        this._complete = complete;
        this.onCompleteChangedEmitter.fire();
    }

    set description(description: string) {
        if (this._description !== description) {
            this._description = description;
            this.onUpdateEmitter.fire();
        }
    }

    get description() {
        return this._description;
    }

    dispose(): void {
        this.toDispose.dispose();
    }
}

export interface WizardDialogProps {
    acceptLabel?: string;
    width?: number;
    height?: number;
}

export class WizardDialog extends BaseDialog<boolean> {
    protected readonly previousBtn: HTMLButtonElement | undefined;
    protected readonly nextBtn: HTMLButtonElement | undefined;
    protected readonly currentPageDisposable = new DisposableCollection();
    protected oldCurrent: WizardPage | undefined;
    private performingFinish = false;
    value = false;

    constructor(protected readonly wizard: Wizard, props?: WizardDialogProps) {
        super({
            title: wizard.title
        });
        const acceptLabel =
            props && props.acceptLabel ? props.acceptLabel : messages.finish;

        if (wizard.hasPrevious) {
            this.previousBtn = this.appendButton(this.createButton(messages.previous));
            this.previousBtn.addEventListener("click", () => this.wizard.previous());
        }
        if (wizard.hasNext) {
            this.nextBtn = this.appendButton(this.createButton(messages.next));
            this.nextBtn.addEventListener("click", () => this.wizard.next());
        }
        this.appendCloseButton(messages.cancel_btn);
        this.appendAcceptButton(acceptLabel);

        if (props && (props.width || props.height)) {
            const style = (this.node.children.item(0) as HTMLDivElement).style;
            style.justifyContent = "flex-start";
            if (props.width) {
                style.maxWidth = `${props.width}px`;
                style.width = `${props.width}px`;
            }

            if (props.height) {
                style.maxHeight = `${props.height}px`;
                style.height = `${props.height}px`;
            }
        }

        this.contentNode.setAttribute(
            "style",
            "justify-content: flex-start; padding-left: 0; padding-right: 0; overflow-y: hidden;"
        );
    }

    protected async init(progress: Progress): Promise<void> {
        if (this.acceptButton)
            this.setEnabled(this.acceptButton, false);
        await this.wizard.init(progress);
        this.wizard.onCompleteChanged(() => this.updateButtons());
        this.wizard.onUpdate(() => this.update());
        this.updateButtons();
        this.update();
    }

    protected doRender(): React.ReactNode {
        this.currentPageDisposable.dispose();
        if (this.wizard.currentPage) {
            this.currentPageDisposable.pushAll([
                this.wizard.currentPage.onCompleteChanged(() => this.updateButtons()),
                this.wizard.currentPage.onUpdate(() => this.update())
            ]);
        }
        return this.wizard.render();
    }

    protected async accept(): Promise<void> {
        if (!this.resolve || this.performingFinish || this.acceptButton?.disabled || !this.wizard.canFinish())
            return;
        try {
            if (this.acceptButton)
                this.setEnabled(this.acceptButton, false);
            this.performingFinish = true;
            this.value = await this.showProgress(undefined, async progress => await this.wizard.performFinish(progress));
        } finally {
            if (this.acceptButton)
                this.setEnabled(this.acceptButton, true);
            this.performingFinish = false;
        }
        if (this.value) {
            this.wizard.dispose();
            return super.accept();
        }
    }

    close(): void {
        this.wizard.dispose();
        super.close();
    }

    protected updateButtons() {
        if (this.isInProgress())
            return;
        if (this.previousBtn) {
            this.setEnabled(
                this.previousBtn,
                this.wizard.getPreviousPage(this.wizard.currentPage) !== undefined
            );
        }

        if (this.nextBtn) {
            this.setEnabled(
                this.nextBtn,
                this.wizard.currentPage !== undefined &&
                this.wizard.currentPage.complete &&
                this.wizard.getNextPage(this.wizard.currentPage) !== undefined
            );
        }
        if (this.acceptButton)
            this.setEnabled(this.acceptButton, this.wizard.canFinish());

        if (this.oldCurrent !== this.wizard.currentPage) {
            this.oldCurrent = this.wizard.currentPage;
            this.update();
        }
    }

    private appendButton(button: HTMLButtonElement): HTMLButtonElement {
        this.controlPanel.appendChild(button);
        button.classList.add("secondary");
        return button;
    }

}
