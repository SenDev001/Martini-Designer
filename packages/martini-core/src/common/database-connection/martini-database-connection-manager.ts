import { JsonRpcServer } from "@theia/core";
import { MartiniEvent } from "../event/martini-event";
import { startCase } from "lodash";

export enum DatabaseType {
    CASSANDRA = "cassandra",
    JDBC = "jdbc",
    MONGODB = "mongodb",
    REDIS = "redis"
}

export interface DatabaseConnection {
    name: string;
    autoStart: boolean;
    status: "STARTED" | "STOPPED";
    type: DatabaseType;
}

export namespace DatabaseConnection {
    export function is(object: any): object is DatabaseConnection {
        return (
            !!object &&
            typeof object === "object" &&
            "name" in object &&
            "autoStart" in object &&
            "status" in object &&
            "type" in object
        );
    }
    export function getType(type: string): DatabaseType | undefined {
        const typeKey: string | undefined = Object.keys(DatabaseType).find(
            (key: string) => DatabaseType[key as keyof typeof DatabaseType] === type
        );
        return typeKey ? DatabaseType[typeKey as keyof typeof DatabaseType] : undefined;
    }
}

export const martiniDatabaseConnectionManagerPath = "/services/martini/database/connections";

export const MartiniDatabaseConnectionManager = Symbol("MartiniDatabaseConnectionManager");

export interface MartiniDatabaseConnectionManager extends JsonRpcServer<MartiniDatabaseConnectionManagerClient> {
    getAll(): Promise<DatabaseConnection[]>;

    get(name: string): Promise<DatabaseConnection>;

    start(name: string): Promise<void>;

    stop(name: string): Promise<void>;

    delete(name: string): Promise<void>;

    save(connection: DatabaseConnection): Promise<void>;

    test(connection: DatabaseConnection): Promise<void>;

    getDrivers(): Promise<Driver[]>;
}

export interface DatabaseConnectionEvent extends MartiniEvent {
    event: "SAVED" | "DELETED" | "STARTED" | "STOPPED" | "ENABLED" | "DISABLED";
    name: string;
}

export namespace DatabaseConnectionEvent {
    export function is(event: MartiniEvent): event is DatabaseConnectionEvent {
        return event.type === "DATABASE_CONNECTION";
    }
}

export interface MartiniDatabaseConnectionManagerClient {
    onEvent(event: DatabaseConnectionEvent): void;
}

export const getTypeLabel = (type: DatabaseType) => {
    if (type === DatabaseType.JDBC) return "JDBC";
    return startCase(type.toLowerCase());
};

export enum JdbcIsolationLevel {
    READ_COMMITTED = "READ_COMMITTED",
    READ_UNCOMMITTED = "READ_UNCOMMITTED",
    REPEATABLE_READ = "REPEATABLE_READ",
    SERIALIZABLE = "SERIALIZABLE"
}

export interface Driver {
    templateUrl: string;
    name: string;
    className: string;
    xaClassName: string;
}

export interface JdbcDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.JDBC;
    url: string;
    xa: boolean;
    driverClassName: string;
    username?: string;
    password: string;
    acquireIncrement: number;
    acquisitionInterval: number;
    acquisitionTimeout: number;
    allowLocalTransactions: boolean;
    applyTransactionTimeout: boolean;
    automaticEnlistingEnabled: boolean;
    deferConnectionRelease: boolean;
    enableJdbc4ConnectionTest: boolean;
    ignoreRecoveryFailures: boolean;
    isolationLevel: JdbcIsolationLevel;
    maxIdleTime: number;
    minPoolSize: number;
    maxPoolSize: number;
    preparedStatementCacheSize: number;
    shareTransactionConnections: boolean;
    testQuery?: string;
    twoPcOrderingPosition: number;
    uniqueName?: string;
    useTmJoin: boolean;
    loginTimeout: number;
}

export enum CassandraCompression {
    NONE = "NONE",
    SNAPPY = "SNAPPY",
    LZ4 = "LZ4"
}

export enum CassandraConsistencyLevel {
    ANY = "ANY",
    ONE = "ONE",
    TWO = "TWO",
    THREE = "THREE",
    QUORUM = "QUORUM",
    ALL = "ALL",
    LOCAL_QUORUM = "LOCAL_QUORUM",
    EACH_QUORUM = "EACH_QUORUM",
    SERIAL = "SERIAL",
    LOCAL_SERIAL = "LOCAL_SERIAL",
    LOCAL_ONE = "LOCAL_ONE"
}

export interface CassandraDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.CASSANDRA;
    contactPoints: string[];
    port: number;
    maxSchemaAgreementWaitSeconds: number;
    credentials: {
        username?: string;
        password: string;
    };
    compression: CassandraCompression;
    metrics: boolean;
    ssl: boolean;
    poolingOptions: {
        coreConnectionsPerHostLocal: number;
        coreConnectionsPerHostRemote: number;
        maxConnectionsPerHostLocal: number;
        maxConnectionsPerHostRemote: number;
        newConnectionThresholdLocal: number;
        newConnectionThresholdRemote: number;
        maxRequestsPerConnectionLocal: number;
        maxRequestsPerConnectionRemote: number;
        idleTimeoutSeconds: number;
        poolTimeoutMillis: number;
        maxQueueSize: number;
        heartbeatIntervalSeconds: number;
    };
    socketOptions: {
        connectTimeoutMillis: number;
        readTimeoutMillis: number;
        keepAlive: boolean;
        reuseAddress: boolean;
        soLinger: number;
        tcpNoDelay: boolean;
        receiveBufferSize: number;
        sendBufferSize: number;
    };
    queryOptions: {
        consistencyLevel: CassandraConsistencyLevel;
        serialConsistencyLevel: CassandraConsistencyLevel;
        fetchSize: number;
        defaultIdempotence: boolean;
        prepareOnAllHosts: boolean;
        reprepareOnUp: boolean;
        refreshSchemaIntervalMillis: number;
        maxPendingRefreshSchemaRequests: number;
        refreshNodeListIntervalMillis: number;
        maxPendingRefreshNodeListRequests: number;
        refreshNodeIntervalMillis: number;
        maxPendingRefreshNodeRequests: number;
    };
}

export enum MongoDbAuthenticationMechanism {
    AUTO = "AUTO",
    PLAIN = "PLAIN",
    SCRAM_SHA_1 = "SCRAM_SHA_1",
    SCRAM_SHA_256 = "SCRAM_SHA_256"
}

export enum MongoDbClusterType {
    STANDALONE = "STANDALONE",
    REPLICA_SET = "REPLICA_SET",
    SHARDED = "SHARDED",
    UNKNOWN = "UNKNOWN"
}

export interface MongoDbDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.MONGODB;
    connectionString: string;
    username?: string;
    password: string;
    authenticationDatabase?: string;
    authenticationMechanism: MongoDbAuthenticationMechanism;
    clusterSettings: {
        localThreshold: number;
        maxWaitQueueSize: number;
        serverSelectionTimeout: number;
        type: MongoDbClusterType;
        hosts: string[];
    };
    connectionPoolSettings: {
        maintenanceFrequency: number;
        maintenanceInitialDelay: number;
        minSize: number;
        maxConnectionIdleTime: number;
        maxConnectionLifeTime: number;
        maxWaitTime: number;
        maxWaitQueueSize: number;
        maxSize: number;
    };
    serverSettings: {
        heartbeatFrequency: number;
        minHeartbeatFrequency: number;
    };
    socketSettings: {
        connectTimeout: number;
        readTimeout: number;
        receiveBufferSize: number;
        sendBufferSize: number;
    };
}

export enum RedisDisconnectedBehavior {
    DEFAULT = "DEFAULT",
    ACCEPT_COMMANDS = "ACCEPT_COMMANDS",
    REJECT_COMMANDS = "REJECT_COMMANDS"
}
export enum RedisSslProvider {
    JDK = "JDK",
    OPENSSL = "OPENSSL"
}
export interface RedisDatabaseConnection extends DatabaseConnection {
    type: DatabaseType.REDIS;
    uri: string;
    password: string;
    clientOptions: {
        pingBeforeActivateConnection: boolean;
        autoReconnect: boolean;
        cancelCommandsOnReconnectFailure: boolean;
        publishOnScheduler: boolean;
        suspendReconnectOnProtocolFailure: boolean;
        requestQueueSize: number;
        disconnectedBehavior: RedisDisconnectedBehavior;
        socketOptions: {
            connectTimeoutMillis: number;
            keepAlive: boolean;
            tcpNoDelay: boolean;
        };
        sslOptions: {
            sslProvider: RedisSslProvider;
            keystoreFile?: string;
            keystorePassword?: string;
            truststoreFile?: string;
            truststorePassword?: string;
            startTls: boolean;
        };
        connectionOptions: {
            autoFlushCommands: boolean;
            connectionTimeoutMillis: number;
        };
        connectionPoolOptions: {
            maxTotal: number;
            maxIdle: number;
            minIdle: number;
            lifo: boolean;
            fairness: boolean;
            maxWaitMillis: number;
            minEvictableIdleTimeMillis: number;
            softMinEvictableIdleTimeMillis: number;
            numTestsPerEvictionRun: number;
            evictorShutdownTimeoutMillis: number;
            testOnCreate: boolean;
            testOnBorrow: boolean;
            testOnReturn: boolean;
            testWhileIdle: boolean;
            timeBetweenEvictionRunsMillis: number;
            evictionPolicyClassName?: string;
            blockWhenExhausted: boolean;
            jmxEnabled: boolean;
            jmxNameBase?: string;
            jmxNamePrefix?: string;
        };
    };
}
