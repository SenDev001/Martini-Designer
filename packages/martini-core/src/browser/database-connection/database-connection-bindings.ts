import { CommandContribution, MenuContribution } from "@theia/core";
import { OpenHandler, WebSocketConnectionProvider, WidgetFactory } from "@theia/core/lib/browser";
import { interfaces } from "inversify";
import {
    MartiniDatabaseConnectionManager,
    martiniDatabaseConnectionManagerPath
} from "../../common/database-connection/martini-database-connection-manager";
import {
    MartiniMongoDbManager,
    martiniMongoDbManagerPath
} from "../../common/database-connection/mongodb/martini-mongodb-manager";
import {
    MartiniDatabaseSchemaProvider,
    martiniDatabaseSchemaProviderPath
} from "../../common/database-connection/schema/martini-database-schema-provider";
import {
    NavigatorContentProviderContribution,
    NavigatorLabelProviderContribution,
    NavigatorOpenHandler
} from "../navigator/martini-navigator-view-widget";
import { WizardContribution } from "../wizard/wizard-contribution";
import {
    DatabaseConnectionCommandContribution,
    DatabaseConnectionMenuContribution
} from "./database-connection-contribution";
import {
    DatabaseConnectionEventDispatcher,
    DatabaseConnectionEventDispatcherClient
} from "./database-connection-event-dispatcher";
import {
    DatabaseConnectionNavigatorContentProvider,
    DatabaseConnectionNavigatorLabelProvider
} from "./database-connection-navigator-contribution";
import {
    DatabaseConnectionEditor,
    DatabaseConnectionEditorNavigatorOpenHandler,
    DatabaseConnectionEditorOpenHandler,
    DatabaseConnectionEditorOptions
} from "./editor/database-connection-editor";
import {
    DatabaseConnectionWizard,
    DatabaseConnectionWizardContribution,
    DatabaseConnectionWizardPage
} from "./wizard/database-connection-wizard";

export const bindDatabaseConnectionBindings = (bind: interfaces.Bind) => {
    bind(DatabaseConnectionEventDispatcherClient)
        .toSelf()
        .inSingletonScope();
    bind(DatabaseConnectionEventDispatcher).toService(DatabaseConnectionEventDispatcherClient);
    bind(MartiniDatabaseConnectionManager)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(
                martiniDatabaseConnectionManagerPath,
                ctx.container.get(DatabaseConnectionEventDispatcherClient)
            );
        })
        .inSingletonScope();
    bind(MartiniDatabaseSchemaProvider)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(martiniDatabaseSchemaProviderPath);
        })
        .inSingletonScope();
    bind(MartiniMongoDbManager).toDynamicValue(ctx => {
        const connection = ctx.container.get(WebSocketConnectionProvider);
        return connection.createProxy(martiniMongoDbManagerPath);
    });
    bind(NavigatorContentProviderContribution).to(DatabaseConnectionNavigatorContentProvider);
    bind(NavigatorLabelProviderContribution).to(DatabaseConnectionNavigatorLabelProvider);
    bind(CommandContribution)
        .to(DatabaseConnectionCommandContribution)
        .inSingletonScope();
    bind(MenuContribution)
        .to(DatabaseConnectionMenuContribution)
        .inSingletonScope();

    bind(WizardContribution)
        .to(DatabaseConnectionWizardContribution)
        .inSingletonScope();
    bind(DatabaseConnectionWizard).toSelf();
    bind(DatabaseConnectionWizardPage).toSelf();
    bind("Factory<DatabaseConnectionWizard>").toAutoFactory(DatabaseConnectionWizard);

    bind(WidgetFactory)
        .toDynamicValue(({ container }) => ({
            id: DatabaseConnectionEditor.FACTORY_ID,
            createWidget: (options: DatabaseConnectionEditorOptions) => {
                const child = container.createChild();
                child.bind(DatabaseConnectionEditorOptions).toConstantValue(options);
                child.bind(DatabaseConnectionEditor).toSelf();
                return child.get(DatabaseConnectionEditor);
            }
        }))
        .inSingletonScope();

    bind(OpenHandler)
        .to(DatabaseConnectionEditorOpenHandler)
        .inSingletonScope();
    bind(NavigatorOpenHandler)
        .to(DatabaseConnectionEditorNavigatorOpenHandler)
        .inSingletonScope();
};
