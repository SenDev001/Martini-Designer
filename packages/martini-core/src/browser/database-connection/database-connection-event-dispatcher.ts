import { Emitter, Event } from "@theia/core";
import { injectable } from "inversify";
import { DatabaseConnectionEvent } from "../../common/database-connection/martini-database-connection-manager";
import { MartiniEndpointManagerClient } from "../../common/endpoint/martini-endpoint-manager";

export const DatabaseConnectionEventDispatcher = Symbol("DatabaseConnectionEventDispatcher");

export interface DatabaseConnectionEventDispatcher {
    onDatabaseConnectionEvent: Event<DatabaseConnectionEvent>;
}

@injectable()
export class DatabaseConnectionEventDispatcherClient
    implements MartiniEndpointManagerClient, DatabaseConnectionEventDispatcher {
    private readonly onDatabaseConnectionEventEmitter = new Emitter<DatabaseConnectionEvent>();
    readonly onDatabaseConnectionEvent = this.onDatabaseConnectionEventEmitter.event;

    onEvent(event: DatabaseConnectionEvent): void {
        this.onDatabaseConnectionEventEmitter.fire(event);
    }
}
