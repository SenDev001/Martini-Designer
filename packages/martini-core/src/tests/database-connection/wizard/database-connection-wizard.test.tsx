import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { noop } from "lodash";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import "reflect-metadata";
import { CreateDatabaseConnectionForm } from "../../../browser/database-connection/wizard/database-connection-wizard";
import { DatabaseType } from "../../../common/database-connection/martini-database-connection-manager";

Enzyme.configure({ adapter: new Adapter() });

test("CreateDatabaseConnectionForm should be rendered", () => {
    const component = Renderer.create(<CreateDatabaseConnectionForm
        onChange={noop}
        onValidate={noop}
        existingNames={[]}
        defaultConfig={{
            type: DatabaseType.CASSANDRA,
            name: "test"
        }}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
