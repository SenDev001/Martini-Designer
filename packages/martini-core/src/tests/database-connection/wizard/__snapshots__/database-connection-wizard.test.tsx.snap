// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDatabaseConnectionForm should be rendered 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
  style={
    Object {
      "display": "grid",
      "gridColumnGap": 10,
      "gridTemplateColumns": "max-content 1fr",
    }
  }
>
  <label
    htmlFor="type"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Type
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr",
        "gridTemplateRows": "max-content 1fr",
      }
    }
  >
    <select
      className="theia-select"
      name="type"
      onBlur={[Function]}
      onChange={[Function]}
      value="cassandra"
    >
      <option
        value="cassandra"
      >
        Cassandra
      </option>
      <option
        value="jdbc"
      >
        JDBC
      </option>
      <option
        value="mongodb"
      >
        Mongodb
      </option>
      <option
        value="redis"
      >
        Redis
      </option>
    </select>
    <div
      className="form-error-message-wrapper"
    />
  </div>
  <label
    htmlFor="name"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Name
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr max-content",
        "gridTemplateRows": "max-content 1fr",
      }
    }
  >
    <input
      autoFocus={true}
      className="theia-input"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value="test"
    />
    <div
      style={
        Object {
          "fontSize": "16px",
          "marginLeft": "4px",
        }
      }
      title="Pressing Up or Down arrow while editing the name changes the type."
    >
      â‡…
    </div>
    <div
      className="form-error-message-wrapper"
    />
  </div>
</form>
`;
